File: apps/server/src/codebot/api/schemas/projects.py
"""Pydantic v2 schemas for project CRUD endpoints."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class ProjectCreate(BaseModel):
    """Request body for creating a new project."""

    name: str = Field(min_length=1, max_length=255)
    description: str = ""
    prd_source: str = "text"
    prd_content: str = ""
    prd_url: str | None = None
    prd_file: str | None = None
    source_name: str | None = None
    source_media_type: str | None = None
    project_type: str | None = None
    repository_path: str | None = None
    repository_url: str | None = None
    tech_stack: dict | None = None
    settings: dict | None = None

    @field_validator("prd_source")
    @classmethod
    def validate_prd_source(cls, value: str) -> str:
        """Normalize and validate the PRD source selector."""
        normalized = value.strip().lower().replace("-", "_")
        allowed = {"text", "url", "file"}
        if normalized not in allowed:
            raise ValueError(f"prd_source must be one of {sorted(allowed)}")
        return normalized

    @field_validator("project_type")
    @classmethod
    def validate_project_type(cls, value: str | None) -> str | None:
        """Normalize optional project type values."""
        if value is None:
            return None
        normalized = value.strip().lower().replace("-", "_")
        allowed = {"greenfield", "inflight", "brownfield", "improve"}
        if normalized not in allowed:
            raise ValueError(f"project_type must be one of {sorted(allowed)}")
        return normalized

    @model_validator(mode="after")
    def validate_source_payload(self) -> "ProjectCreate":
        """Ensure the selected PRD source has the required payload."""
        if self.prd_source == "url" and not (self.prd_url or "").strip():
            raise ValueError("prd_url is required when prd_source is 'url'")
        if self.prd_source == "file" and not (self.prd_file or "").strip():
            raise ValueError("prd_file is required when prd_source is 'file'")
        return self


class ProjectUpdate(BaseModel):
    """Request body for updating a project (partial)."""

    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None


class ProjectResponse(BaseModel):
    """Project data returned in API responses."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    description: str
    status: str
    project_type: str
    prd_format: str
    tech_stack: dict | None
    created_at: datetime
    updated_at: datetime

    @field_validator("status", "project_type", mode="before")
    @classmethod
    def enum_to_string(cls, v: object) -> str:
        """Convert ORM enum to lowercase string."""
        if hasattr(v, "value"):
            return str(v.value).lower()
        return str(v).lower()


class ProjectDetailResponse(ProjectResponse):
    """Extended project data including PRD content and config."""

    prd_content: str
    config: dict | None
    repository_path: str
    repository_url: str | None

File: apps/dashboard/src/api/projects.ts
import { api } from "./client";
import type { Project } from "@/types/project";

export interface ProjectCreatePayload {
  name: string;
  description: string;
  prd_source: "text" | "file" | "url";
  prd_content?: string;
  prd_url?: string;
  prd_file?: string;
  source_name?: string;
  source_media_type?: string;
  project_type?: "greenfield" | "inflight" | "brownfield" | "improve";
  repository_path?: string;
  repository_url?: string;
  tech_stack?: Record<string, unknown>;
  settings?: Record<string, unknown>;
}

interface ResponseEnvelope<T> {
  status: "success";
  data: T;
  meta: { request_id: string; timestamp: string };
}

interface PaginatedEnvelope<T> {
  status: "success";
  data: T[];
  meta: { request_id: string; timestamp: string };
  pagination: {
    page: number;
    per_page: number;
    total: number;
    total_pages: number;
  } | null;
}

export const projectApi = {
  list(page = 1, perPage = 20) {
    return api.get<PaginatedEnvelope<Project>>(
      `/api/v1/projects?page=${String(page)}&per_page=${String(perPage)}`,
    );
  },

  get(projectId: string) {
    return api.get<ResponseEnvelope<Project>>(
      `/api/v1/projects/${projectId}`,
    );
  },

  create(data: ProjectCreatePayload) {
    return api.post<ResponseEnvelope<Project>>("/api/v1/projects", data);
  },

  update(projectId: string, data: Partial<Pick<Project, "name" | "description">>) {
    return api.patch<ResponseEnvelope<Project>>(
      `/api/v1/projects/${projectId}`,
      data,
    );
  },

  delete(projectId: string) {
    return api.delete<void>(`/api/v1/projects/${projectId}`);
  },
};
